cpp/src/join/key_remapping.cu

```diff
@@ -0,0 +1,641 @@
+                                               cuda::thread_scope_device,
+                                               BuildEqual,
+                                               probing_scheme_type,
+                                               rmm::mr::polymorphic_allocator<char>,
+                                               cuco_storage_type>;
+
+  key_remap_table()                                  = delete;
+  ~key_remap_table() override                        = default;
+  key_remap_table(key_remap_table const&)            = delete;
+  key_remap_table(key_remap_table&&)                 = default;
+  key_remap_table& operator=(key_remap_table const&) = delete;
+  key_remap_table& operator=(key_remap_table&&)      = default;
+
+  /**
+   * @brief Construct a key remap hash table with counting.
+   */
+  template <typename RowHasher>
+  key_remap_table(
+    cudf::table_view const& build,
+    std::shared_ptr<cudf::detail::row::equality::preprocessed_table> preprocessed_build,
+    BuildEqual const& build_equal,
+    RowHasher const& row_hasher,
+    cudf::null_equality compare_nulls,
+    double load_factor,
+    rmm::cuda_stream_view stream)
+    : _has_nested_columns{cudf::has_nested_columns(build)},
+      _compare_nulls{compare_nulls},
+      _build{build},
+      _preprocessed_build{std::move(preprocessed_build)},
+      _hash_table{cuco::extent{static_cast<std::size_t>(build.num_rows())},
+                  load_factor,
+                  cuco::empty_key{cuco::pair{std::numeric_limits<hash_value_type>::max(),
+                                             rhs_index_type{cudf::JoinNoMatch}}},
+                  build_equal,
+                  {},
+                  cuco::thread_scope_device,
+                  cuco_storage_type{},
+                  rmm::mr::polymorphic_allocator<char>{},
+                  stream.value()},
+      _distinct_count{0},
+      _max_duplicate_count{0}
+  {
+    CUDF_FUNC_RANGE();
+    CUDF_EXPECTS(0 != this->_build.num_columns(), "Key remap build table is empty");
+
+    cudf::size_type const build_num_rows{_build.num_rows()};
+    if (build_num_rows == 0) { return; }
+
+    // Allocate counts array - one entry per build row
+    rmm::device_uvector<cudf::size_type> counts(build_num_rows, stream);
+    thrust::fill(rmm::exec_policy_nosync(stream), counts.begin(), counts.end(), 0);
+
+    // Device scalar for distinct count
+    rmm::device_scalar<cudf::size_type> d_distinct_count(0, stream);
+
+    // Get hash table reference for insert_and_find
+    auto set_ref = _hash_table.ref(cuco::op::insert_and_find);
+
+    // Create the key pair iterator
+    auto const key_iter = cudf::detail::make_counting_transform_iterator(
+      0, make_key_pair<rhs_index_type, RowHasher>{row_hasher});
+
+    // Determine if we need to filter nulls
+    bool const skip_nulls =
+      (_compare_nulls == cudf::null_equality::UNEQUAL) && cudf::nullable(build);
```

Comment:
Use `table_view::has_nested_nulls` for checking nested columns.

Reply:
But that is not what Spark does, and it will not work properly for lists. I don't think that null equality is applied equally in all cases.